_posts/2023-09-07-develop-functions-locally.md

```diff
@@ -0,0 +1,343 @@
+
+## Generated image tags
+
+All OpenFaaS functions are built into container images. By default if no image tag is included for a function in the stack.yml file the `:latest` tag is used. When iterating over functions and pushing them to an image registry it is a best practice to organise different image versions using tags instead of always pushing to `:latest`.
+
+There are two options to set tags for function images.
+
+1. Let the faas-cli generate the tag automatically.
+2. Set the image tag in the stack.yml file.
+
+The `--tag` option can be used with the `build`, `push` and `deploy` sub-commands of the faas-cli. If this flag is provided, image tags for functions will automatically be generated based on available metadata. This can be either Git metadata like the commit sha or branch name or digest of the function handler content.
+
+The generated tag is always suffixed to any tag defined in the stack.yml file or `latest` if no tag is defined.
+
+Some examples:
+
+When using the flag `--tag=sha` the image tag used in the stack.yml file is suffixed with the short Git SHA. e.g
+
+```yaml
+functions:
+  echo:
+    lang: node18
+    handler: ./echo
+    image: echo:0.2
+```
+
+For this stack.yml file the resulting image name will be `echo:0.2-cf59cfc`
+
+If no tag is set in the stack.yml file the suffix is appended to latest.
+
+```
+image: echo => image: echo:latest-cf59cfc      
+```
+
+If you are using `faas-cli up` with the `--watch` flag we recommend also setting `--tag=digest`. The digest is calculated from the function source code and will ensure a unique image tag is generated and pushed for every code change.
+
+> Find an overview of [all the available tag versions](https://docs.openfaas.com/cli/tags/) in the docs.
+
+Alternatively [environment variable substitution](https://docs.openfaas.com/reference/yaml/#yaml-environment-variable-substitution) can be used to set the image.
+
+Here is an example of a stack.yml file:
+
+```yaml
+image: my-fn:${FN_VERSION:-latest}
+```
+
+The value of `FN_VERSION` can be set through en environment variable when running commands like `faas-cli up`, `build` or `publish`: 
+```
+FN_VERSION="0.2" faas-cli build
+```
+
+Environment variable substitution and the `--tag` flag can also be used together.
+
+## Conclusion
+We set out to show you how the new `local-run` command worked for faster iteration, the `--watch` flag for live-reloading as you edit, and the `--tag=digest` flag to generate dynamic tags as you edit.
+
+When it comes to iterating on functions, doing a full deployment is the easiest and most realistic way to test your work, however, faas-cli local-run can speed things up if your function doesn't have a lot of dependencies.
+
+For local development you can use a local registry with KinD to save on bandwidth and speed up the development cycle when using `faas-cli up`: [Use a local registry with KinD](https://docs.openfaas.com/tutorials/local-kind-registry/)
```

Comment:
This doesn't seem to have good context, since we've been explaining what makes local development quicker, then we offer an alternative - I think that's going to confuse people.

I would suggest working it in higher up in the article, next to where you show how to use `faas-cli up --watch` - and explain that it has trade-offs, like some additional work.